src/parser.rs

```diff
@@ -1,377 +1,1440 @@
-    message<Message>,
-    map!(
-        message_events,
-        |(name, events): (String, Vec<MessageEvent>)| {
-            let mut msg = Message {
-                name: name.clone(),
-                ..Message::default()
+trait ToI64 {
+    fn to_i64(&self) -> ParserResult<i64>;
+}
+
+trait ToChar {
+    fn to_char(&self) -> ParserResult<char>;
+}
+
+impl ToI32 for u64 {
+    fn to_i32(&self) -> ParserResult<i32> {
+        if *self <= i32::max_value() as u64 {
+            Ok(*self as i32)
+        } else {
+            Err(ParserError::IntegerOverflow)
+        }
+    }
+}
+
+impl ToI32 for i64 {
+    fn to_i32(&self) -> ParserResult<i32> {
+        if *self <= i32::max_value() as i64 && *self >= i32::min_value() as i64 {
+            Ok(*self as i32)
+        } else {
+            Err(ParserError::IntegerOverflow)
+        }
+    }
+}
+
+impl ToI64 for u64 {
+    fn to_i64(&self) -> Result<i64, ParserError> {
+        if *self <= i64::max_value() as u64 {
+            Ok(*self as i64)
+        } else {
+            Err(ParserError::IntegerOverflow)
+        }
+    }
+}
+
+impl ToChar for u8 {
+    fn to_char(&self) -> Result<char, ParserError> {
+        if *self <= 0x7f {
+            Ok(*self as char)
+        } else {
+            Err(ParserError::ExpectAscii)
+        }
+    }
+}
+
+impl ToU8 for u32 {
+    fn to_u8(&self) -> Result<u8, ParserError> {
+        if *self as u8 as u32 == *self {
+            Ok(*self as u8)
+        } else {
+            Err(ParserError::IntegerOverflow)
+        }
+    }
+}
+
+
+#[derive(Clone)]
+pub struct Parser<'a> {
+    input: &'a str,
+    pos: usize,
+    syntax: Syntax,
+    pub line: u32,
+    pub col: u32,
+}
+
+#[derive(Default)]
+pub struct MessageBody {
+    pub fields: Vec<Field>,
+    pub oneofs: Vec<OneOf>,
+    pub reserved_nums: Vec<FieldNumberRange>,
+    pub reserved_names: Vec<String>,
+    pub messages: Vec<Message>,
+    pub enums: Vec<Enumeration>,
+}
+
+impl<'a> Parser<'a> {
+    pub fn new(input: &'a str) -> Parser<'a> {
+        Parser {
+            input,
+            pos: 0,
+            syntax: Syntax::Proto2,
+            line: FIRST_LINE,
+            col: FIRST_COL,
+        }
+    }
+
+    fn rem_chars(&self) -> &'a str {
+        &self.input[self.pos..]
+    }
+
+    /// No more chars
+    fn lexer_eof(&self) -> bool {
+        self.rem_chars().is_empty()
+    }
+
+    /// No more tokens
+    fn syntax_eof(&self) -> ParserResult<bool> {
+        let mut clone = self.clone();
+        clone.skip_ws()?;
+        Ok(clone.lexer_eof())
+    }
+
+    // higher level functions
+
+    fn parser_opt<R, F>(&mut self, f: F) -> Option<R>
+        where F : FnOnce(&mut Parser) -> Option<R>
+    {
+        let mut clone = self.clone();
+        let r = f(&mut clone);
+        if let Some(_) = r {
+            *self = clone;
+        }
+        r
+    }
+
+    /// Apply a parser and return a string which matched
+    fn parser_source<F>(&mut self, f: F) -> Option<&'a str>
```

Comment:
It's called `source` [in jparsec library](http://jparsec.github.io/jparsec/apidocs/org/codehaus/jparsec/Parser.html#source--), so I chose this name. Couldn't find quickly what's the parser name in the original Haskell parsec library, but nevermind, I'll simply inline the function since it's used only in one place.